dbcon/joblist/tupleaggregatestep.cpp

```diff
@@ -5728,116 +5742,147 @@ uint64_t TupleAggregateStep::doThreadedAggregate(ByteStream& bs, RowGroupDL* dlp
+              // The distinctAggregator accumulates the aggregation results of all row groups by being added
+              // all row groups of each bucket aggregator and doing an aggregation step after each addition.
+              auto* bucketMultiDistinctAggregator = dynamic_cast<RowAggregationMultiDistinct*>(fAggregators[bucketNum].get());
+              auto* bucketDistinctAggregator = dynamic_cast<RowAggregationDistinct*>(fAggregators[bucketNum].get());
+              distinctAggregator->aggregator(bucketDistinctAggregator->aggregator());
 
-                if (aggMultiDist)
-                {
-                  (dynamic_cast<RowAggregationMultiDistinct*>(agg))
-                      ->subAggregators(aggMultiDist->subAggregators());
-                }
-
-                agg->doDistinctAggregation();
-              }
-              // for "group by without distinct" case
-              else
+              if (bucketMultiDistinctAggregator)
               {
-                fAggregator->append(fAggregators[i].get());
+                (dynamic_cast<RowAggregationMultiDistinct*>(distinctAggregator))
+                    ->subAggregators(bucketMultiDistinctAggregator->subAggregators());
               }
+
+              distinctAggregator->doDistinctAggregation();
             }
           }
         }
 
+        // Deliver results
         fDoneAggregate = true;
+        bool done = true;
+        while (fAggregator->nextRowGroup() && !cancelled())
+        {
+          done = false;
+          fAggregator->finalize();
+          rowCount = fRowGroupOut.getRowCount();
+          fRowsReturned += rowCount;
+          fRowGroupDelivered.setData(fRowGroupOut.getRGData());
+
+          if (rowCount != 0)
+          {
+            if (!cleanUpAndOutputRowGroup(bs, dlp))
+              break;
+          }
+          done = true;
+        }
+        if (done)
+          fEndOfResult = true;
+      }
+    } else if (hasGroupByColumns) {
+      // CASE 3: Query contains no aggregation on a DISTINCT column, but at least one GROUP BY column
+      // e.g. SELECT SUM(col1) FROM test GROUP BY col2;
+      // As the aggregation algorithm of first phase calculates aggregation results online, e.g. for SUM()
+      // the aggregation here is already finished and the row groups just need to be delivered.
+      // TODO: Previous code meant that no more aggregation steps are done here. Is this always the case?
+
+      if (!fEndOfResult)
+      {
+        if (!fDoneAggregate)
+        {
+          // Do aggregation over all row groups. As all row groups need to be aggregated together there is no
+          // easy way of multi-threading this and it's done in a single thread for now.
+          for (uint32_t bucketNum = 0; bucketNum < fNumOfBuckets; bucketNum++)
+          {
+            if (fEndOfResult == false)
+            {
+              fAggregator->append(fAggregators[bucketNum].get());
+            }
+          }
+        }
       }
```

Comment:
looks the whole block could be simplified to
````
else if (hasGroupByColumns && !fEndOfResult && !fDoneAggregate) 
{
    for (uint32_t bucketNum = 0; bucketNum < fNumOfBuckets; bucketNum++)
    {
         fAggregator->append(fAggregators[bucketNum].get());
    }
}
```
please apply clang-format as well, curly braces style is inconsistent here